kroxylicious-kms-provider-aws-kms/DEV_GUIDE.md

```diff
@@ -0,0 +1,81 @@
+Here's some information that might be helpful when developing this module.
+
+# Testing the ability to authenticate with EC2 metadata
+
+In order to test this feature, you need to deploy Kroxylicious to an EC2 instance. The 
+EC2 instance needs to be associated with a properly configured IAM role. The EC2 credentials
+provider will use that to authenticate.
+
+Here's some instructions that will help you test end to end.  For simplicity, these instructions
+co-locate a Kafka Broker on the same instance.  Of course, you can choose to run the Kafka
+Cluster in any network accessible location.  These instructions also assume you want to run
+your Kafka client off-EC2.
+
+1. Create an EC2 instance on AWS - `t2.micro` instance will suffice.
+2. Enable SSH with key pair. The remainder of the instructions assume you've download the key into `.ssh/MyEC2KeyPair.pem`.
+3. Enable inbound traffic to port 22 (for SSH) and 9192 - 9193 (for Kafka traffic).
+4. Follow the 'Authenticating using AWS EC2 metadata' instructions in https://kroxylicious.io/kroxylicious/#assembly-aws-kms-proxy
+   to create the policy, IAM role and assign the IAM role to the EC2 host.
+5. SSH into the EC2 instance `ssh  -i "~/.ssh/MyEC2KeyPair.pem" ec2-xxx-xxx-xxx-xxx.compute-1.amazonaws.com` and install packages ready to run docker containers and kroxylicious.
+   ```bash
+    sudo yum install -y java maven docker
+    sudo service docker start
+    sudo usermod -a -G docker ec2-user
+   ```
+6. Logout/login to EC2 again to pick up the new group.
+7. Run Kafka on the EC2 instance (inside a docker container)
+   ```bash
+   docker run -d -p 9092:9092 --name broker apache/kafka:latest
+   ```
+8. Scp a Kroxylicious dist to the EC2 instance.
+   ```bash
+   scp  -i "~/.ssh/MyEC2KeyPair.pem" /kroxylicious-app-0.9.0-SNAPSHOT-bin.zip ec2-user@ec2-xxx-xxx-xxx-xxx.compute-1.amazonaws.com:.
+   ```
+9. Back on the EC2 instance, unzip Kroxylicious
+   ```bash
+   unzip kroxylicious-app-0.9.0-SNAPSHOT-bin.zip
+   ```
+10. Create a config file for Kroxylicious updating the virtual clusters bootstrap address to the public address of the EC2 instance.
+    ```yaml
+     virtualClusters:
+     - demo:
+       targetCluster:
+         bootstrap_servers: localhost:9092
+       clusterNetworkAddressConfigProvider:
+         type: PortPerBrokerClusterNetworkAddressConfigProvider
+         config:
+           bootstrapAddress: ec2-xx-xx-xx-xx.compute-1.amazonaws.com:9192
+       logNetwork: false
+       logFrames: false
+     filters:
+     - type: RecordEncryption
+       config:
+         kms: AwsKmsService
+         kmsConfig:
+           region: us-east-1
+           endpointUrl: https://kms.us-east-1.amazonaws.com
+         credentialsProvider:
```

Comment:
this needs to be indented further, under kmsConfig. Also this pointed to something that could be improved.

When I had it configured like this it failed with a `NPE` inside the fixed credentials provider, which is a bit surprising. Maybe something should complain earlier if neither `credentialsProvider` or the legacy settings are present.

Reply:
fixed.  the configuration mechanism will change anyway.